python-checks/src/main/java/org/sonar/python/checks/OverwrittenCollectionEntryCheck.java

```diff
@@ -0,0 +1,196 @@
+        reportOverwrites(ctx, collectionWrites);
+        collectionWrites.clear();
+      }
+    }
+    reportOverwrites(ctx, collectionWrites);
+  }
+
+  private static Expression lhs(AssignmentStatement assignment) {
+    return assignment.lhsExpressions().get(0).expressions().get(0);
+  }
+
+  @CheckForNull
+  private static CollectionWrite collectionWrite(AssignmentStatement assignment, Expression expression) {
+    if (expression.is(Kind.SLICE_EXPR)) {
+      SliceExpression sliceExpression = (SliceExpression) expression;
+      String key = key(sliceExpression.sliceList().slices());
+      return collectionWrite(assignment, sliceExpression.object(), key, sliceExpression.leftBracket(), sliceExpression.rightBracket());
+
+    } else if (expression.is(Kind.SUBSCRIPTION)) {
+      SubscriptionExpression subscription = (SubscriptionExpression) expression;
+      String key = key(subscription.subscripts().expressions());
+      return collectionWrite(assignment, subscription.object(), key, subscription.leftBracket(), subscription.rightBracket());
+
+    } else {
+      return null;
+    }
+  }
+
+  private static CollectionWrite collectionWrite(AssignmentStatement assignment, Expression collection, @Nullable String key, Token lBracket, Token rBracket) {
+    if (key == null) {
+      return null;
+    }
+
+    if (collection.is(Kind.SLICE_EXPR, Kind.SUBSCRIPTION)) {
+      CollectionWrite nested = collectionWrite(assignment, collection);
+      if (nested != null) {
+        return new CollectionWrite(nested.collectionKey.nest(key), nested.leftBracket, rBracket, assignment);
+      }
+    }
+
+    if (collection instanceof HasSymbol) {
+      Symbol symbol = ((HasSymbol) collection).symbol();
+      if (symbol != null) {
+        CollectionKey collectionKey = new CollectionKey(symbol, key);
+        return new CollectionWrite(collectionKey, lBracket, rBracket, assignment);
+      }
+    }
+
+    return null;
+  }
+
+  @CheckForNull
+  private static String key(List<? extends Tree> subscriptsOrSlices) {
+    StringBuilder key = new StringBuilder();
+    for (Tree tree : subscriptsOrSlices) {
+      String keyElement = key(tree);
+      if (keyElement == null) {
+        return null;
+      }
+      key.append(",").append(keyElement);
+    }
+    return key.toString();
+  }
+
+  private static String key(Tree tree) {
+    if (tree.is(Kind.NUMERIC_LITERAL)) {
+      return ((NumericLiteral) tree).valueAsString();
+    } else if (tree.is(Kind.STRING_LITERAL)) {
+      return ((StringLiteral) tree).trimmedQuotesValue();
+    } else if (tree.is(Kind.NAME)) {
+      return ((Name) tree).name();
+    } else if (tree.is(Kind.SLICE_ITEM)) {
+      SliceItem sliceItem = (SliceItem) tree;
+      List<String> keyParts = Stream.of(sliceItem.lowerBound(), sliceItem.upperBound(), sliceItem.stride())
+        .map(e -> e == null ? "" : key(e))
+        .collect(Collectors.toList());
+      return keyParts.contains(null) ? null : String.join(":", keyParts);
```

Comment:
Can `keyParts` actually contain null here?
From the code it doesn't look so and there are no test case fulfilling this condition.